### ISSUE TITLE
No support for IMEs

### ISSUE THREAD

**Issue by greggman** [capnmidnight/Primrose#26]
If you're Japanese, Chinese, Korean, Thai, etc your S.O.L. at the moment. I don't know if there's a reasonable solution but google sheets handles this case as an example an they are also rendering to a canvas.

![google sheets](http://i.imgur.com/xxvFPBh.png)

**Comment by capnmidnight** [comment]
Hi, thanks for the report. I do have a scaffolding in place for handling this sort of thing, but being as the project is only about a month old, I haven't quite gotten around to figuring out exactly what data is sent for such things.

But I suspect it will just be figuring out the right sequences of key codes and creating code pages out of them: https://github.com/capnmidnight/Primrose/tree/master/src/code_pages

You wouldn't happen to know off of the top of your head where one could find good documentation for this, would you? It would save me some trouble trying to track it down.

**Comment by greggman** [comment]
I don't think it's that simple, though maybe I don't understand code pages. 

IMEs pop up their own tiny window(s) wherever the insertion point is so in general I'm guessing you'd need to put a text box at the closest 2d place to your 3d display and then wait for the IME to give you whatever text the user entered. 

Excuse me if you already know all of this but if not here's a gif showing Japanese input. 

![ime](http://i.imgur.com/JmmYwyi.gif)

Those windows and completions appearing are displayed by the OS not the app. Basically the as long as the words are underlined the OS/IME is in control, all the app sees is the words once the underline disappears (the user has pressed enter in the IME). In other words while the underline appears the user is in the IME. They type stuff, the IME lets them select what that stuff will become, they then press enter and the IME sends the user's selection to the app. But to do that requires using the OS's input system vs reading the keyboard directly.

A native app can call IME api methods to help with better integration. A web page though is limited to using an `<input type="text"` or `<textarea>` or a `contenteditable` to allow the IME to appear AFAIK.

In Primrose's case the IME appears in the top left (because there's no insertion point?) and when the user presses enter to choose their completion that is not then inserted in the Primrose document. 

Maybe there's just some message you need to handle to get that input from the browser.

Of course it's not really that simple because IME's support *reconversion* (control-shift-R on OSX). This is selecting some text and then giving it back to the IME so the user can change what the conversion was. There's no way to do that from JavaScript either. You have to use a textarea/input so the browser can give the text back to the IME.

![reconversion](http://i.imgur.com/ThkGCSR.gif)

**Comment by capnmidnight** [comment]
I've been banging this around in my head for the last month. 

There is some effort to provide programmatic access to handle events related to IMEs in a standard way: http://www.w3.org/TR/ime-api But as far as I can tell, there is no implementation of it.

I have recently added some code to make a hidden text area that echos the behavior of Primrose. See issues #52, #53, and #54. Originally, this was to make the cursor and text completion work correctly on mobile devices. It should also help to make the IME appear in the right place, but only in 2D use cases. 

In 3D use cases, I still need to implement issue #51 to be able to get the hidden text area to be in the right place so the OS will know where to put the IME. I don't know if that will actually work, I fear the browser may not consider CSS3D when working on this, but it's about the only thing that can be done.

To actually get the converted text to show up, I'd have to tear apart the key event handling code. I'd have to let that hidden text area take control of actual text editing, which means I'm stuck to how the browser wants to present text and will never be able to implement features like Sublime Text's multi-headed cursor. That's not the end of the world, however.

I'm not going to speculate on performance issues. I've been surprised on several occasions where the code has been spending the most time. Thankfully, the Chrome profiler is pretty good.

But there is one, final issue that I'm afraid might punt this feature off into the far future: in VR, the IME is only going to appear in one eye, and it's going to be the wrong size and location for viewing comfortably. And that was kind of the whole point of this project in the first place: current 2D controls don't work in 3D, stereo environments. 

Long story short, the only solution that will work for all issues is to completely implement a new IME in JavaScript. That, unfortunately, will have to wait until after I've developed more controls for VR.

**Comment by capnmidnight** [comment]
And implementing an IME is not as crazy as it sounds, because supporting VR headsets with no physical attachment to a PC means I'll have to create a software keyboard anyway and use hand gestures to activate it. It's all work that has to be done. Ideally, this would all be done at the window manager level, but I sincerely doubt that's going to happen any time soon.

**Comment by capnmidnight** [comment]
Trello card: https://trello.com/c/0qYPIV9o
